scheduler/reserve.go

```diff
@@ -0,0 +1,128 @@
+package scheduler
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/Sirupsen/logrus"
+	"time"
+)
+
+// ReserveAction is an interface to provide a way to reserve a resource in scheduler
+type ReserveAction interface {
+	Reserve(scheduler *Scheduler, requests []ResourceRequest, context Context, host *host, force bool, data map[string]interface{}) error
+	RollBack(scheduler *Scheduler, requests []ResourceRequest, context Context, host *host)
+}
+
+// ProvideReserveAction provides a list of reserve action. It is a gateway to register a new reserve action
+func ProvideReserveAction() []ReserveAction {
+	actions := []ReserveAction{}
+
+	// add ComputeReserveAction
+	actions = append(actions, &ComputeReserveAction{})
+
+	// add PortReserveAction
+	actions = append(actions, &PortReserveAction{})
+
+	return actions
+}
+
+// ComputeReserveAction is a reserve action for cpu, memory, instance count and other compute resource.
+type ComputeReserveAction struct {
+	offset int
+}
+
+func (c *ComputeReserveAction) Reserve(scheduler *Scheduler, requests []ResourceRequest, context Context, host *host, force bool, data map[string]interface{}) error {
+	var err error
+	var reserveLog *bytes.Buffer
+	for _, rr := range requests {
+		p, ok := host.pools[rr.GetResourceType()]
+		if !ok {
+			logrus.Warnf("Pool %v for host %v not found for reserving %v. Skipping reservation", rr.GetResourceType(), host.id, rr)
+			continue
+		}
+		PoolType := p.GetPoolType()
+		if PoolType == computePool {
+			pool := p.(*ComputeResourcePool)
+			request := rr.(AmountBasedResourceRequest)
+			if !force && pool.Used+request.Amount > pool.Total {
+				err = OverReserveError{hostID: host.id, resourceRequest: rr}
+				return err
+			}
+
+			pool.Used = pool.Used + request.Amount
+			c.offset = c.offset + 1
+			if reserveLog == nil {
+				reserveLog = bytes.NewBufferString(fmt.Sprintf("New pool amount on host %v:", host.id))
+			}
+			reserveLog.WriteString(fmt.Sprintf(" %v total: %v used: %v.", request.Resource, pool.Total, pool.Used))
+		}
+	}
+	if reserveLog != nil {
+		logrus.Info(reserveLog.String())
+	}
+	return nil
+}
+
+func (c *ComputeReserveAction) RollBack(scheduler *Scheduler, requests []ResourceRequest, context Context, host *host) {
+	time.Sleep(time.Second * 2)
```

Comment:
legacy debug code.... i removed it in the next two commits so this commit is still showing it...